From c9edb73496d7c9e7d62a8c71c2668cc19db0529a Mon Sep 17 00:00:00 2001
From: Yongzhi Chen <ychena@apache.org>
Date: Fri, 12 May 2017 10:06:04 -0400
Subject: [PATCH 1121/1164] CDH-58947 HIVE-16660: Not able to add partition
 for views in hive when sentry is enabled (Yongzhi
 Chen, reviewed by Aihua Xu)

Change-Id: Id27f02fd1623b2ae7003c85d470297a106da9c07
---
 .../apache/hadoop/hive/ql/metadata/Partition.java  |   10 ++++++++--
 .../hadoop/hive/ql/parse/DDLSemanticAnalyzer.java  |    4 +++-
 2 files changed, 11 insertions(+), 3 deletions(-)

diff --git a/ql/src/java/org/apache/hadoop/hive/ql/metadata/Partition.java b/ql/src/java/org/apache/hadoop/hive/ql/metadata/Partition.java
index 074dae1..66d5070 100644
--- a/ql/src/java/org/apache/hadoop/hive/ql/metadata/Partition.java
+++ b/ql/src/java/org/apache/hadoop/hive/ql/metadata/Partition.java
@@ -239,9 +239,15 @@ public Path getPartitionPath() {
 
   public Path getDataLocation() {
     if (table.isPartitioned()) {
-      return new Path(tPartition.getSd().getLocation());
+      if (tPartition.getSd() == null)
+        return null;
+      else
+        return new Path(tPartition.getSd().getLocation());
     } else {
-      return new Path(table.getTTable().getSd().getLocation());
+      if (table.getTTable() == null || table.getTTable().getSd() == null)
+        return null;
+      else
+        return new Path(table.getTTable().getSd().getLocation());
     }
   }
 
diff --git a/ql/src/java/org/apache/hadoop/hive/ql/parse/DDLSemanticAnalyzer.java b/ql/src/java/org/apache/hadoop/hive/ql/parse/DDLSemanticAnalyzer.java
index 4b3b098..949eea2 100644
--- a/ql/src/java/org/apache/hadoop/hive/ql/parse/DDLSemanticAnalyzer.java
+++ b/ql/src/java/org/apache/hadoop/hive/ql/parse/DDLSemanticAnalyzer.java
@@ -2819,7 +2819,9 @@ private void analyzeAlterTableAddParts(String[] qualified, CommonTree ast, boole
         }
         cmd.append(")");
       }
-      Driver driver = new Driver(conf);
+      SessionState ss = SessionState.get();
+      String uName = (ss == null? null: ss.getUserName());
+      Driver driver = new Driver(conf, uName);
       int rc = driver.compile(cmd.toString(), false);
       if (rc != 0) {
         throw new SemanticException(ErrorMsg.NO_VALID_PARTN.getMsg());
-- 
1.7.9.5

